dbms/src/DataStreams/IBlockInputStream.h

```diff
@@ -314,7 +314,11 @@ class IBlockInputStream
         /// NOTE: Acquire a read lock, therefore f() should be thread safe
         std::shared_lock lock(children_mutex);
 
-        for (auto & child : children)
+        // Reduce lock scope and avoid recursive locking since that is undefined for shared_mutex.
+        const auto children_copy = children;
```

Comment:
Since we do not remove any entry from `children`, another option would be to use a `std::list` and memorizing the `children.end()` to eliminate copying of std::vector.

A draft:
```
// IBlockInputStream attribute:
std::list<BlockInputStreamPtr> children;

// ...
    template <typename F>
    void forEachChild(F && f)
    {
        std::shared_lock lock(children_mutex);
        const auto end = children.end();
        lock.unlock();

        for (auto child = children.begin(); child != end; ++child)
            if (f(*child))
                return;
    }
```

Reply:
Usually we don't have many children (notable exception is Merge table) and copying is not very expensive (it involves copying of shared_ptrs) and happens infrequently. This issue will become irrelevant after switching to processors. So, not worth to change to list.